test/FinalIOSpec.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE RecursiveDo #-}
+module FinalIOSpec where
+
+import Test.Hspec
+
+import Data.Either
+import Data.IORef
+
+import Polysemy
+
+import Polysemy.State
+import Polysemy.Trace
+import Polysemy.Final.Async
+import Polysemy.Final.Fixpoint
+import Polysemy.Final.Error
+
+data Node a = Node a (IORef (Node a))
+
+mkNode :: (Member (Lift IO) r, Member Fixpoint r)
+       => a
+       -> Sem r (Node a)
+mkNode a = mdo
+  let nd = Node a p
+  p <- sendM $ newIORef nd
+  return nd
+
+linkNode :: Member (Lift IO) r
+         => Node a
+         -> Node a
+         -> Sem r ()
+linkNode (Node _ r) b =
+  sendM $ writeIORef r b
+
+readNode :: Node a -> a
+readNode (Node a _) = a
+
+follow :: Member (Lift IO) r
+       => Node a
+       -> Sem r (Node a)
+follow (Node _ ref) = sendM $ readIORef ref
+
+test1 :: IO (Either Int (String, Int, Maybe Int))
+test1 = do
+  ref <- newIORef "abra"
+  runFinal
+    . runStateInIORef ref -- Order of these interpreters don't matter
+    . runErrorInIOFinal
+    . runFixpointFinal
+    . runAsyncFinal
```

Comment:
This is an example of a scenario where you can't use `runAsync`. For some reason, the combination of exceptions and fixpoint results in it being blocked indefinitely on a `MVar`. (It doesn't matter if you replace the other `-Final` interpreters with their non-`Final `counterparts.)